### PR TITLE
[rng] Initial implement sceKernelGetRandomNumber

### PR DESCRIPTION
**Pull request by d3m3vilurr**
only tested linux system.

### CODE REVIEW

File: src/emulator/modules/SceLibKernel/SceLibRng.cpp [hunk 1]
```diff
@@ -17,8 +17,56 @@
...
+
+void read_random(void *output, unsigned int size) {
+    getrandom(output, size, GRND_RANDOM);
+}
+
+#endif
+
+EXPORT(int, sceKernelGetRandomNumber, void *output, unsigned int size) {
+    if (size > 64) {
+        return -1;
```
Comment (Rinnegatamante) [inline src/emulator/modules/SceLibKernel/SceLibRng.cpp]: This should be replaced with proper erroring using RET_ERROR

File: src/emulator/modules/SceLibKernel/SceLibRng.cpp [hunk 2]
```diff
@@ -17,8 +17,56 @@
...
+    if (size != (DWORD)size) {
+        return;
+    }
+
+    if (CryptAcquireContext(&csp, NULL, NULL, PROV_RSA_FULL,
+                            CRYPT_VERIFYCONTEXT) == false) {
+        return;
+    }
+
+    if (CryptGenRandom(csp, (DWORD)size, buffer) == false) {
```
Comment (Rinnegatamante) [inline src/emulator/modules/SceLibKernel/SceLibRng.cpp]: If possible, avoid C casts and prefer static_cast.

File: src/emulator/modules/SceLibKernel/SceLibRng.cpp (hunk 1 above)
Comment (VelocityRa) [reply src/emulator/modules/SceLibKernel/SceLibRng.cpp]: use `error`, my PR with `RET_ERROR` isn't merged yet

**Comment by d3m3vilurr** [comment]
done

File: src/emulator/modules/SceLibKernel/SceLibRng.cpp [hunk 3]
```diff
@@ -16,9 +16,40 @@
...
+#if defined(WIN32)
+#include <windows.h>
+#include <wincrypt.h>
+#else
+#include <fstream>
+#endif
 
-EXPORT(int, sceKernelGetRandomNumber) {
-    return unimplemented("sceKernelGetRandomNumber");
+#define SCE_RNG_ERROR_INVALID_ARGUMENT  0x810C0000
```
Comment (VelocityRa) [inline src/emulator/modules/SceLibKernel/SceLibRng.cpp]: do not align values

**Comment by d3m3vilurr** [comment]
because this api fill random buffer instead of generating a random number. and I want to avoid seed setting like `srand`.
(and I don't know c++ std not very well ;))
also, probably this function wrapped the bsd's [read_random](https://www.freebsd.org/cgi/man.cgi?query=read_random&apropos=0&sektion=9&manpath=FreeBSD+11-current&format=html)
it also used `/dev/random` for the secure reason.

File: src/emulator/modules/SceLibKernel/SceLibRng.cpp [hunk 4]
```diff
@@ -16,9 +16,33 @@
...
+#include <algorithm>
+#include <random>
 
-EXPORT(int, sceKernelGetRandomNumber) {
-    return unimplemented("sceKernelGetRandomNumber");
+#define SCE_RNG_ERROR_INVALID_ARGUMENT 0x810C0000
+
+EXPORT(int, sceKernelGetRandomNumber, uint64_t *output, unsigned int size) {
+    if (size > 64) {
+        return error("sceKernelGetRandomNumber", SCE_RNG_ERROR_INVALID_ARGUMENT);
```
Comment (VelocityRa) [inline src/emulator/modules/SceLibKernel/SceLibRng.cpp]: use `RET_ERROR`

File: src/emulator/modules/SceLibKernel/SceLibRng.cpp [hunk 5]
```diff
@@ -16,9 +16,33 @@
...
+    thread_local std::mt19937_64 mt(dev());
+    thread_local auto rng = [&]() { return mt(); };
+
+    int repeat = size / sizeof(uint64_t);
+    int remain = size % sizeof(uint64_t);
+
+    std::generate_n(output, repeat, rng);
+
+    if (remain == 0) {
+        return 0;
```
Comment (VelocityRa) [inline src/emulator/modules/SceLibKernel/SceLibRng.cpp]: `return SCE_KERNEL_OK`

File: src/emulator/modules/SceLibKernel/SceLibRng.cpp [hunk 6]
```diff
@@ -16,9 +16,33 @@
...
+    }
+
+    uint64_t value = rng();
+    char *ptr = reinterpret_cast<char*>(&value);
+    std::copy(ptr, ptr + remain, reinterpret_cast<char*>(&output[repeat]));
+    return 0;
```
Comment (VelocityRa) [inline src/emulator/modules/SceLibKernel/SceLibRng.cpp]: `return SCE_KERNEL_OK`

File: src/emulator/modules/SceLibKernel/SceLibRng.cpp [hunk 7]
```diff
@@ -16,9 +16,33 @@
 // 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
 #include "SceLibRng.h"
+#include <algorithm>
```
Comment (VelocityRa) [inline src/emulator/modules/SceLibKernel/SceLibRng.cpp]: add newline before this

**Comment by d3m3vilurr** [comment]
done & rebased

**Pull request merged**